# ПРОЕКТ ПО SQL 

## Описание проекта

**Цель исследования** - получить данные для формулировки ценностного предложения для крупного сервиса для чтения книг по подписке.

**Задача** - проанализировать базу данных (информация о книгах, издательствах, авторах, а также пользовательские обзоры книг).

**Шаги для анализа базы данных:**

- Посчитать, сколько книг вышло после 1 января 2000 года;

- Для каждой книги посчитать количество обзоров и среднюю оценку;

- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;

- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;

- Посчитайть среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [72]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [73]:
query = ''' SELECT *
        FROM books ''' # таблица с данными о книгах

pd.io.sql.read_sql(query, con = engine) # выводим результат

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

<div class="alert alert-info">
    
**Комментарий студента**
    
База содержит информацию о 1000 книгах (идентификатор, название, кол-во страниц,дата публикации и идентификатор издательства).
    
</div>

In [74]:
query = ''' SELECT *
        FROM authors ''' # таблица с данными об авторах

pd.io.sql.read_sql(query, con = engine) # выводим результат

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

<div class="alert alert-info">
    
**Комментарий студента**
    
База содержит информацию о 636 авторах (их идентификатор и имя).
    
</div>

In [75]:
query = ''' SELECT *
        FROM publishers ''' # таблица с данными об издательствах

pd.io.sql.read_sql(query, con = engine) # выводим результат

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

<div class="alert alert-info">
    
**Комментарий студента**
    
База содержит информацию о 340 издательства (их идентификатор и название).
    
</div>

In [76]:
query = ''' SELECT *
        FROM ratings ''' # таблица с данными о пользовательских оценках книг

pd.io.sql.read_sql(query, con = engine) # выводим результат

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

<div class="alert alert-info">
    
**Комментарий студента**
    
База содержит информацию о 3456 пользовательскх оценках книг (идентификатор оценки, идентификатор книги, имя пользователя и оценка книги).
    
</div>

In [77]:
query = ''' SELECT *
        FROM reviews ''' # таблица с данными о пользовательских обзорах на книги

pd.io.sql.read_sql(query, con = engine) # выводим результат

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<div class="alert alert-info">
    
**Комментарий студента**
    
База содержит информацию о 2793 пользовательских обзорах на книги (идентификатор обзора, идентификатор книги, имя пользователя и текст обзора).
    
</div>

- Посчитать, сколько книг вышло после 1 января 2000 года;

In [78]:
query = ''' SELECT COUNT(book_id)
        FROM books
        WHERE publication_date > '2000-01-01'  '''

pd.io.sql.read_sql(query, con = engine) # выводим результат

count
0    819

<div class="alert alert-info">
    
**Комментарий студента**
    
В базе 819 книг, которые вышли после 1 января 2000 года. 
    
</div>

- Для каждой книги посчитать количество обзоров и среднюю оценку;

In [79]:
query = ''' WITH r AS (SELECT book_id,
                              COUNT(review_id) AS review_count
                       FROM reviews
                       GROUP BY book_id)
            SELECT b.title,
                   review_count,
                   AVG(rat.rating)
            FROM books AS b
            JOIN r ON b.book_id = r.book_id
            JOIN ratings AS rat ON b.book_id = rat.book_id
            GROUP BY b.title, 
                 review_count
            ORDER BY AVG(rat.rating) DESC '''


pd.io.sql.read_sql(query, con = engine) # выводим результат

title  review_count   avg
0                            Triptych (Will Trent  #1)             2  5.00
1    Crucial Conversations: Tools for Talking When ...             2  5.00
2                        A Dirty Job (Grim Reaper  #1)             4  5.00
3                         Plum Island (John Corey  #1)             2  5.00
4                      Act of Treason (Mitch Rapp  #9)             2  5.00
..                                                 ...           ...   ...
989  The World Is Flat: A Brief History of the Twen...             3  2.25
990                  His Excellency: George Washington             2  2.00
991                                              Junky             2  2.00
992                                      Drowning Ruth             3  2.00
993                               Harvesting the Heart             2  1.50

[994 rows x 3 columns]

<div class="alert alert-info">
    
**Комментарий студента**
    
Для каждой книги посчитаны количество обзоров и средняя оценка. Таблица отсортирована по средней оценке. Следует отметить, что в базе есть книги без обзоров (NaN).
    
</div>

- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;

In [80]:
query = ''' SELECT p.publisher,
                   COUNT(book_id)
            FROM books AS b
            JOIN publishers AS p ON b.publisher_id = p.publisher_id
            WHERE num_pages > 50
            GROUP BY p.publisher
            ORDER BY COUNT(book_id) DESC
            LIMIT 1 '''

pd.io.sql.read_sql(query, con = engine) # выводим результат

publisher  count
0  Penguin Books     42

<div class="alert alert-info">
    
**Комментарий студента**
    
Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books	(42 книги).
    
</div>

- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;

In [81]:
query = ''' WITH rat AS (SELECT book_id,
                                AVG(rating) AS avg_rating,
                                COUNT(rating_id) AS count_rating
                         FROM ratings
                         GROUP BY book_id)
            SELECT a.author,
                   AVG(rat.avg_rating) AS avg 
            FROM books AS b
            JOIN rat ON b.book_id = rat.book_id
            JOIN authors AS a ON b.author_id = a.author_id
            WHERE rat.count_rating >= 50
            GROUP BY a.author
            ORDER BY avg DESC
            LIMIT 1 '''

pd.io.sql.read_sql(query, con = engine) # выводим результат

author       avg
0  J.K. Rowling/Mary GrandPré  4.283844

<div class="alert alert-info">
    
**Комментарий студента**
    
Автор с самой высокой средней оценкой книг — J.K. Rowling/Mary GrandPré	 (учитывались книги с 50 и более оценками).
    
</div>

- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [82]:
query = ''' WITH rat AS (SELECT username,
                                COUNT(rating_id) AS count_rating
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(rating_id) > 50), 
            rev AS (SELECT username,
                           COUNT(review_id) AS count_review
                    FROM reviews
                    GROUP BY username)
            SELECT ROUND(AVG(count_review)) as avg_review
            FROM rat
            JOIN rev ON rat.username = rev.username '''

pd.io.sql.read_sql(query, con = engine) # выводим результат

avg_review
0        24.0

<div class="alert alert-info">
    
**Комментарий студента**
    
Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24. 
    
</div>

<div class="alert alert-info">
    
**Комментарий студента**

**Общий вывод**

Анализ базы данных крупного сервиса для чтения книг по подписке показал, что:
- В базе 819 книг, которые вышли после 1 января 2000 года.
- Для каждой книги посчитаны количество обзоров и средняя оценка. Таблица отсортирована по средней оценке. Следует отметить, что в базе есть книги без обзоров (NaN)
- Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books (42 книги).
- Автор с самой высокой средней оценкой книг — J.K. Rowling/Mary GrandPré (учитывались книги с 50 и более оценками).
- Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24. 
    
</div>